### ISSUE TITLE
Potential infinite loop within ParentProcessUtils.GetSelfAndAncestors()

### ISSUE THREAD

**Issue by piers7** [approvals/ApprovalTests.Net#158]
TL/DR: I believe the code in ParentProcessUtils.GetSelfAndAncestors() needs a guard to prevent infinite looping in the case that Win32 calls to determine parent process describe a circular process chain (cause unclear).
## 

After my tests started timing out after 4 hours I had to take some memory dumps to work out what was going on. I noticed that in both cases, my testing process (MSBuild) was sat with VisualStudioReporter.GetParentProcess() as the top(ish) stack frame, and I had a _very_ high (300k) number of ProcessInfo objects in the CLR heaps. In one case I left it another 10 mins before taking another another dump, and the process was on the same stack frame, but with another 20k ProcessInfo objects on the heap.

I'm using ApprovalTests nuget 3.0.7.

Looking in the github source (accepting this may have diverged slightly from what I'm running) it seems like what's happened is:
- the generic DiffReporter has failed to find Beyond Compare etc... on my build server, and pumped for the VisualStudioReporter
- the VisualStudio reporter static ctor is attempting to work out if it was launched from Visual Studio
- this is attempting to recursively walk process IDs and find parent processes (via some Win32 interop calls)
- based on the memory dump and CLR heap this appears to be stuck in an infinite loop
- this appears to only be possible if the Win32 calls are describing a circular parent/process reference chain

This problem was repeatable across test runs (ie processes), so I'm thinking there may have been something wacky either in my parent process (TeamCity)'s memory, or possibly a system wide memory corruption in Win32 somehow. After a system reboot the tests all ran fine again.

I'm raising this issue because if this can happen, even in crazy edge-cases, I'm thinking there should be a max recursion guard in that code, because it seems like fail-quiet would be a better bet in this case (spending 4 hours in a static class constructor seems like overkill :). A simple loopup of existing 'visited' process IDs would suffice.

(If anyone's really interested I'll attach sections of the stack traces, but I can't really attach the whole dump because passwords etc...)

**Comment by jamesrcounts** [comment]
I wrote the visual studio reporter and wish I never had.  Having said that I'm in favor of any fail silent approach.  It is far more common for people to use a real diff utility, and not the poor excuse offered by VS (without TFS, although the one you get with TFS is pretty terrible too).  

But the frontloaded reporter should have detected TeamCity before any of that happened.  And of course we're on 3.0.11 now so its natural to wonder whether upgrading will help.

**Comment by piers7** [comment]
Ok, so I'm not using the FrontLoadedReporter, because I didn't know what it was, but having looked at this http://stackoverflow.com/questions/9939209/how-to-use-approvaltests-on-teamcity then I suspect I can indeed bypass the problem entirely by doing all that.

Also, looking again at the code on github, the recursion scenario does now seems to be explicitly handled in the code so presumably 3.0.11 would indeed help.
https://github.com/approvals/ApprovalTests.Net/blob/master/ApprovalTests/Utilities/ParentProcessUtils.cs
```
                    // loop detected (parent id have been re-allocated to a child process!)
                    yield break;
```

**Comment by SimonCropp** [comment]
yep @piers7 is correct. this should now be fixed
